ChatQnA/docker/chatqna_no_wrapper.py

```diff
@@ -0,0 +1,226 @@
+        del inputs["text"]
+    return inputs
+
+
+def align_outputs(self, data, cur_node, inputs, runtime_graph, llm_parameters_dict):
+    next_data = {}
+    if self.services[cur_node].service_type == ServiceType.EMBEDDING:
+        assert isinstance(data, list)
+        next_data = {"text": inputs["inputs"], "embedding": data[0]}
+    elif self.services[cur_node].service_type == ServiceType.RETRIEVER:
+
+        docs = [doc["text"] for doc in data["retrieved_docs"]]
+
+        with_rerank = runtime_graph.downstream(cur_node)[0].startswith("rerank")
+        if with_rerank and docs:
+            # forward to rerank
+            # prepare inputs for rerank
+            # TODO add top_n
+            next_data["query"] = data["initial_query"]
+            next_data["texts"] = [doc["text"] for doc in data["retrieved_docs"]]
+        else:
+            # forward to llm
+            if not docs:
+                # delete the rerank from retriever -> rerank -> llm
+                for ds in reversed(runtime_graph.downstream(cur_node)):
+                    for nds in runtime_graph.downstream(ds):
+                        runtime_graph.add_edge(cur_node, nds)
+                    runtime_graph.delete_node_if_exists(ds)
+
+            # prepare inputs for LLM
+            next_data["parameters"] = {}
+            next_data["parameters"].update(llm_parameters_dict)
+            del next_data["parameters"]["id"]
+            del next_data["parameters"]["chat_template"]
+            del next_data["parameters"]["streaming"]
+            prompt = ChatTemplate.generate_rag_prompt(data["initial_query"], docs)
+            next_data["inputs"] = prompt
+
+    elif self.services[cur_node].service_type == ServiceType.RERANK:
+        # prepare inputs for LLM
+        next_data["parameters"] = {}
+        next_data["parameters"].update(llm_parameters_dict)
+        del next_data["parameters"]["id"]
+        del next_data["parameters"]["chat_template"]
+        del next_data["parameters"]["streaming"]
+
+        # rerank the inputs with the scores
+        # TODO add top_n
+        top_n = 1
+        docs = inputs["texts"]
+        reranked_docs = []
+        for best_response in data[:top_n]:
+            reranked_docs.append(docs[best_response["index"]])
+
+        prompt = ChatTemplate.generate_rag_prompt(inputs["query"], reranked_docs)
+        next_data["inputs"] = prompt
+
+    return next_data
+
+
+def align_generator(self, gen):
+    # TGI format
+    # {"index":20,"token":{"id":368,"text":" you","logprob":0.0,"special":false},"generated_text":null,"details":null}
+    for line in gen:
+        line = line.decode("utf-8")
+        start = line.find("{")
+        end = line.rfind("}") + 1
+
+        json_str = line[start:end]
+        try:
+            # sometimes yield empty chunk, do a fallback here
+            json_data = json.loads(json_str)
+            yield f"data: {repr(json_data['token']['text'].encode('utf-8'))}\n\n"
```

Comment:
should we follow openai respone format here?

Reply:
Does our UI accept openai compatible response? I've tested it with our current UI and this line makes the output correctly displayed on frontend.